gpii/node_modules/preferencesServer/src/preferencesService.js

```diff
@@ -223,6 +486,47 @@ var fluid = fluid || require("infusion");
         return promiseTogo;
     };
 
+    /**
+     * Stand alone method to add a `gpiiKey` document in the system. `prefsSafeId` and `prefSetId` are
+     * required fields. Optionally you can specify the `gpiiKey` otherwise a new one will be generated.
+     * This key must not exist yet.
+     * @param {Object} that - Instance of `gpii.preferencesServer.preferencesService`.
+     * @param {Object} dataStore - Instance of `gpii.dbOperation.dataStore`
+     * @param {Object} keyData - Data for the new `gpiiKey` document.
+     * @return {Promise} A promise object that resolves to the new saved document, or rejects with a
+     * standard error payload.
+     */
+    gpii.preferencesServer.preferencesService.addGpiiKey = function (that, dataStore, keyData) {
+        var promTogo = fluid.promise();
+
+        if (!keyData.prefsSafeId || !keyData.prefsSetId) {
+            promTogo.reject({
+                isError: true,
+                message: "Unable to add GPII Key to Preferences Safe. `prefsSafeId` and `prefsSetId` are required fields."
+            });
+        }
+        else {
+            dataStore.addGpiiKey({
```

Comment:
Adding a key without verifying the existence of prefsSafeId and prefsSetId will result in the unsynced data in "gpiiKey" and "prefsSafe" documents.

Reply:
Good call, I added a check for this, and unit tests for it.